colm.1
.\"@(#)colm.1	1.1 6/3/88 22:48:54
.RL "MASSCOMP"
.TH LS 1
.UC
.SH NAME
colm \- columnate lines of text
.SH SYNOPSIS
.B colm
[
.B  -cglnstvwLT
] name ...
.SH DESCRIPTION
.I colm
produces multi-column output from single-column input.
It is sort of a cross between
.I pr
-<n> (which prints multi-column output sorted across)
and the column-sorting function
of
.I ls
(1) (which sorts down, but only if given no filename arguments).
It can produce fixed or variable width columns, and it figures out any
options not explicitly specified.
.PP
.I colm
uses getopt(3C) to parse its options, so options may be strung
together, as in
.IR -vw ,
and values may either adhere to their option letters or follow as
the next argument.
The options are as follows:
.TP
.B \-v
Produce variable-width columns.
.I colm
uses an iterative algorithm to find the maximum number of columns that
allows at least the gutter-width between the widest element and the next
column.
The -v option is incompatible with -c.
.TP
.BI \-c n
Use fixed-width columns of width
.IR n .
Data will be truncated if necessary to fit the specified column width.
If
.B -c
is not used,
.I colm
expands the column width(s) to fit the output width.
.TP
.BI \-g n
Use a "gutter" of
.I n
characters between columns.
The default is 1.
.TP
.BI \-n n
Use
.I n
fixed-width columns.
Without
.BR -n ,
.I colm
fits as many columns in the given output width as it can.
.TP
.B -s
Spread the columns to fit the output width.
Otherwise columns are made as narrow as possible,
while leaving the gutter space between columns.
With fixed-width columns, the columns will be spread as far as
possible while keeping their widths even.
WIth variable-width columns, the columns will be spread so that there
is an even amount of white space between them.
.TP
.BI \-w n
Sets the output width to
.I n
characters.
The output lines may be shorter, but cannot be longer than
.IR n .
.TP
.BI \-l n
Sets the page length to
.I n
lines.
This option is not implemented.
An error message will result if you try to use it.
.TP
.BI \-L c
Sets the
.I leader character
to
.IR c .
This character is used repeatedly to fill in the space between columns.
The default is to use as many spaces and tabs as are needed.
-L ' ' will cause
.I colm
not to use tabs.
Watch out for shell metacharacters.
.TP
.BI \-t s
Use the literal string "s" to separate the columns,
rather than a repeated leader character.
Watch out for shell metacharacters.
.br
There are two main differences between -L and -t:
-L only allows a single character, while -t allows a string;
and the leader character specified by -L is repeated to fill the
space, whereas the string specified with -t is used only once between
columns.
Generally, -t will not produce even columns, while -L always will.
.TP
.BI \-T n
Specifies the output tab width as
.I n
characters.
The default is eight.
.SH DIAGNOSTICS
Several errors and warnings can be produced by
.IR colm :
.br
A usage message is produced for illegal options.
.br
If any single input line is longer than the output width,
.I colm
terminates with an error.
.br
Illegal combinations of arguments and impossible conditions,
such as 20 columns of 30 characters
in an output width of 79 columns, or selecting spreading with a fixed
separator string, cause errors.
.br
Certain combinations of total number of lines and specified number of
columns cannot produce balanced columns, even with the last column
allowed to be short.
If this occurs,
.I colm
prints an explanatory message on stderr and uses a smaller number
of columns than were specified.
.br
Attempting to do multi-page columnation (with -l) produces an error.
.br
Nonexistent or unreadable files produce an error.
.SH SEE ALSO
pr(1), ls(1)
.SH BUGS
Multi-page columnation is not implemented.
.XX "Text manipulation"
.XX "Columns, creating"


colm.c
/*----------------------------------------------------------------------
 * colm: auto-columnate, with various options.
 * by Gary Oberbrunner  5-31-88
 *
 * -v		variable-width columns (default: fixed-width)
 * -s		spread output columns to fit output width (default: don't)
 * -g #		Separate columns by a #-column gutter
 * -c #		column width (size) (default: max line length of input)
 * -n #		use # columns (default: program picks max # that will fit)
 * -w #		output width is # characters/line (default: 80)
 * -l #		page length is # lines (default: infinite)
 * -L c		leader character is 'c' (default: space)
 * -t s		use string "s" as separator; (default: enough space to line up)
 * -T #		tab size # chars (default 8)
 *
 * Copyright c. 1988, Gary Oberbrunner.
 * This program may be freely redistributed providing that this source code
 * and the accompanying copyright notice are preserved intact.
 *----------------------------------------------------------------------
 */

#include <stdio.h>
#include <string.h>
#include "std.h"
#include "colm.h"

int debug = 1;
static char SCCSId[] = "@(#)colm.c	1.2 6/3/88 (MASSCOMP) 23:17:12";
static char Copyright[] = "Copyright c. 1988, Gary Oberbrunner.";

OPTS Opts;			/* global copy of original options */
main (argc, argv)
int argc;
char **argv;
{
    int	c;
    extern int optind;
    extern char	*optarg;
    int errflg = 0;
    int totlines = 0;

    /* Command-line options: */
    OPTS opts;
    opts.Vflg = FALSE;		/* variable col width */
    opts.Sflg = FALSE;		/* stretch output columns to fit */
    opts.Gutter = 1;		/* gutter width */
    opts.ColWidth = 0;		/* fixed col width */
    opts.Ncols = 0;		/* number of cols */
    opts.Width = 79;		/* output width */
    opts.Tabsize = 8;		/* tab size */
    opts.PageLen = 0;		/* page length */
    opts.ColSep[0] = '\0';	/* column separator string */
    opts.Leader = '\0';		/* leader char for column separation */

    while ((c =	getopt (argc, argv, "vsg:c:n:w:l:L:t:T:")) != EOF) {
	switch	(c) {
	  case 'v':
	    if (opts.ColWidth != 0) errflg++;
	    opts.Vflg = TRUE;
	    break;
	  case 's':
	    opts.Sflg = TRUE;
	    break;
	  case 'c':
	    if (opts.Vflg == TRUE) errflg++;
	    opts.ColWidth = atoi(optarg);
	    if (opts.ColWidth < 1) errflg++;
	    break;
	  case 'g':
	    opts.Gutter = atoi(optarg);
	    if (opts.Gutter < 0) errflg++;
	    break;
	  case 'n':
	    opts.Ncols = atoi(optarg);
	    if (opts.Ncols < 0) errflg++;
	    break;
	  case 'w':
	    opts.Width = atoi(optarg);
	    if (opts.Width < 1) errflg++;
	    break;
	  case 'l':
	    opts.PageLen = atoi(optarg);
	    if (opts.PageLen < 1) errflg++;
	    break;
	  case 'T':
	    opts.Tabsize = atoi(optarg);
	    if (opts.Tabsize < 1) errflg++;
	    break;
	  case 'L':
	    opts.Leader = optarg[0];
	    break;
	  case 't':
	    strcpy(opts.ColSep, optarg);
	    break;
	  case '?':
	    errflg++;
	}
    }
    Opts = opts;		/* copy to global, so we know what was */
				/* actually user-specified and what we */
				/* computed */
    if (errflg)	{
	Usage();
	exit (2);
    }
    if (optind >= argc)
	totlines = ReadIn("-", &Lines); /* read stdin */
    else {
	for( ; optind < argc; optind++) {
	    totlines += ReadIn(argv[optind], &Lines); /* read text files */
	}
    }
    if (totlines > 0)
	Columnate(&Lines, &opts, totlines);
}

Usage()
{
    fprintf(stderr,"\n\
Usage: colm [-cglnstvwLT] <files>\n\
 Use '-' for stdin.\n\
 Options:\n\
 -v	variable-width columns (default: fixed-width)\n\
 -s	spread columns to fit output width (default: don't spread)\n\
 -g #	separate columns by a #-character 'gutter'.\n\
 -c #	column width (size) (default: max line length of input)\n\
 -n #	use # columns (default: program picks max # that will fit)\n\
 -w #	output width is # characters/line (default: 80)\n\
 -l #	page length is # lines (default: infinite)\n\
 -L c	leader character is 'c' (default: space)\n\
 -t s	use string \"s\" as separator (default: enough space to line up)\n\
 -T #	use #-space tabs (default: 8)\n");
}
